contracts/proxy/contract.go

```diff
@@ -42,8 +74,67 @@ func Update(nefFile, manifest []byte, data any) {
 
 // Verify checks whether carrier transaction contains either (2/3N + 1) or
 // (N/2 + 1) valid multi-signature of the NeoFS Alphabet.
+// Container contract's `SubmitObjectPut` is an exception, Alphabet signature
+// is not required, contract's `VerifyPlacementSignatures` is called instead.
 func Verify() bool {
-	return common.ContainsAlphabetWitness()
+	cnrH := storage.Get(storage.GetReadOnlyContext(), containerContractKey).(interop.Hash160)
+	tx := runtime.GetScriptContainer()
+	if string(tx.Script[len(tx.Script)-5:]) != string([]byte{65, 98, 125, 91, 82}) { // SystemContractCall
+		return common.ContainsAlphabetWitness()
+	}
+	scriptNoSyscall := tx.Script[:len(tx.Script)-5]
+	scriptLen := len(scriptNoSyscall)
+	if !cnrH.Equals(scriptNoSyscall[scriptLen-20:]) {
+		return common.ContainsAlphabetWitness()
+	}
+
+	scriptNoContract := scriptNoSyscall[:scriptLen-20-1-1]
+	scriptLen = len(scriptNoContract)
+	expMethod := "submitObjectPut"
+	if scriptLen < len(expMethod) {
+		return common.ContainsAlphabetWitness()
+	}
+	if string(scriptNoContract[scriptLen-len(expMethod):]) != expMethod {
+		return common.ContainsAlphabetWitness()
+	}
+
+	scriptArgsOnly := scriptNoContract[:scriptLen-len(expMethod)-1-1]
```

Comment:
Frankly, this backward reading is not very friendly, iterating in natural order is easier to understand.

Reply:
i feel like a fast fail if it is not a required contract/method/contract syscall was a better choice. also i feel this order is more natural ("i call this contract, this method...")